File: iOS Integration Guide/example-apps/objc-demo-app/DemoApp/ViewController.m
#import "ViewController.h"
#import "PlacedOrderEvent.h"

@interface ViewController ()

@end

@implementation ViewController

- (void) viewDidLoad {
    [super viewDidLoad];
    [Optimove.shared registerSuccessStateDelegate:self];
    
    // Optipush Only
    [Optimove.shared registerWithDeepLinkResponder: [[OptimoveDeepLinkResponder alloc] init: self]];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    // Make sure to release the reference to the state listener and avoid memory leaks
    [Optimove.shared unregisterSuccessStateDelegate:self];
}

// MARK: - Optimove SDK Initialization

- (void)optimove:(Optimove *)optimove didBecomeActiveWithMissingPermissions:(NSArray<NSNumber *> *)missingPermissions {
    // Example for how to handle Optimove initialization
    self.isOptimoveInitialized = true;
    
    for (int i = 0; i < missingPermissions.count; i++) {
        NSNumber *missingPermission = missingPermissions[i];
        if (missingPermission == OptimoveDeviceRequirementAdvertisingId) {
            // Maybe prompt the user to enable IDFA report
        } else if (missingPermission == OptimoveDeviceRequirementAdvertisingId) {
            // Maybe prompt the user to enable Push Notifications
        }
    }
    
    // Make sure to call the Optimove SDK only after it has been initialized
    
    // Report screen visit like this
    [Optimove.shared setScreenVisitWithScreenPath:@"Home/Store/Footwear/Boots" screenTitle:@"<YOUR_TITLE>" screenCategory:@"<OPTIONAL: YOUR_CATEGORY>"];
    // OR
    [Optimove.shared setScreenVisitWithScreenPathArray:@[@"Home", @"Store", @"Footwear", @"Boots"] screenTitle:@"<YOUR_TITLE>" screenCategory:@"<OPTIONAL: YOUR_CATEGORY>"];
}

// Mark - Optimove SDK Indentification

- (void) loginWithEmail: (NSString *)email {
    // Some login logic through which the SDK ID is retrieved
    NSString *sdkId = [MyLoginService getSdkIdWithEmail: email];
    
    if (self.isOptimoveInitialized) {
        if (sdkId != nil) {
            // If the Optimove SDK is initialized AND both the email and the sdkId are valid, you can call the registerUser
            [Optimove.shared registerUserWithSdkId:sdkId email:email];
        } else {
            // If the Optimove SDK is initialized AND only the email is valid, you can call the setUserEmail
            [Optimove.shared setUserEmailWithEmail:email];
        }
    } else {
        // No need to wait for the SDK to be initialized when calling setUserId
        [Optimove.shared setUserId:sdkId];
    }
}


// Mark - Optimove SDK Events

- (void) reportSimpleEvent {
    if (!self.isOptimoveInitialized) { return; }
    [Optimove.shared reportEventWithName:@"signup" parameters: @{@"first_name": @"John",
                                                                 @"last_name": @"Doe",
                                                                 @"email": @"john@doe.com",
                                                                 @"age": @(42),
                                                                 @"opt_in": @(false)}];
}


- (void) reportComplexEvent {
    if (!self.isOptimoveInitialized) { return; }
    [Optimove.shared reportEvent:[[PlacedOrderEvent alloc] initWithCartItems: @[[[CartItem alloc] init]]]];
}


// Mark - Optimove SDK Optipush

- (void)didReceiveWithDeepLink:(OptimoveDeepLinkComponents *)deepLink {
    if (deepLink == nil) { return; }
    
    // Get the targetted screen's name
    NSString *screenName = deepLink.screenName;
    
    // Get the dynamic deep link's params
    NSDictionary *params = deepLink.parameters;
}

- (void) startOptipushTestMode {
    [Optimove.shared startTestMode];
}

- (void) stopOptipushTestMode {
    [Optimove.shared stopTestMode];
}

@end

File: iOS Integration Guide/example-apps/swift-demo-app/DemoApp/ViewController.swift
import UIKit
import OptimoveSDK

class ViewController: UIViewController {
    
    private var isOptimoveInitialized: Bool = false
    
    override func viewDidLoad() {
        super.viewDidLoad()
        Optimove.shared.registerSuccessStateListener(self)
        
        // Optipush Only
        Optimove.shared.register(deepLinkResponder: OptimoveDeepLinkResponder(self))
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        // Make sure to release the reference to the state listener and avoid memory leaks
        Optimove.shared.unregisterSuccessStateListener(self)
    }
}

// MARK: - Optimove SDK Initialization

extension ViewController: OptimoveSuccessStateListener {
    
    func optimove(_ optimove: Optimove, didBecomeActiveWithMissingPermissions missingPermissions: [OptimoveDeviceRequirement]) {
        // Example for how to handle Optimove initialization
        self.isOptimoveInitialized = true
        
        for missingPermission in missingPermissions {
            if missingPermission == .advertisingId {
                // Maybe prompt the user to enable IDFA report
            } else if missingPermission == .userNotification {
                // Maybe prompt the user to enable Push Notifications
            }
        }
        
        // Make sure to call the Optimove SDK only after it has been initialized
        
        // Report screen visit like this
        Optimove.shared.setScreenVisit(screenPath: "Home/Store/Footwear/Boots", screenTitle: "<YOUR_TITLE>", screenCategory: "<OPTIONAL: YOUR_CATEGORY>")
        // OR
        Optimove.shared.setScreenVisit(screenPathArray: ["Home", "Store", "Footwear", "Boots"], screenTitle: "<YOUR_TITLE>", screenCategory: "<OPTIONAL: YOUR_CATEGORY>")
    }
}

// Mark - Optimove SDK Indentification

extension ViewController {
    
    func login(email: String) {
        // Some login logic through which the SDK ID is retrieved
        let sdkId = MyLoginService.getSdkId(email)
        
        
        if self.isOptimoveInitialized {
            if sdkId != nil {
                // If the Optimove SDK is initialized AND both the email and the sdkId are valid, you can call the registerUser
                Optimove.shared.registerUser(sdkId: sdkId, email: email)
            } else {
                // If the Optimove SDK is initialized AND only the email is valid, you can call the setUserEmail
                Optimove.shared.setUserEmail(email: email)
            }
        } else {
            // No need to wait for the SDK to be initialized when calling setUserId
            Optimove.shared.setUserId(sdkId)
        }
    }
}

// Mark - Optimove SDK Events

extension ViewController {
    
    func reportSimpleEvent() {
        if !self.isOptimoveInitialized { return }
        Optimove.shared.reportEvent(name: "signup", parameters: [
            "first_name": "John",
            "last_name": "Doe",
            "email": "john@doe.com",
            "age": 42,
            "opt_in": false
        ])
    }
    
    func reportComplexEvent() {
        if !self.isOptimoveInitialized { return }
        Optimove.shared.reportEvent(PlacedOrderEvent([CartItem]()))
    }
}

// Mark - Optimove SDK Optipush

extension ViewController: OptimoveDeepLinkCallback {

    func didReceive(deepLink: OptimoveDeepLinkComponents?) {
        guard let deepLink = deepLink else {return}
        guard let vc = storyboard?.instantiateViewController(withIdentifier: "deepLinkVc") as? DeepLinkViewController else { return }
        vc.deepLinkComp = deepLink
        present(vc, animated: true)
    }
    
    @IBAction func startOptipushTestMode(_ sender: UIButton) {
        Optimove.shared.startTestMode()
    }
    
    @IBAction func stopOptipushTestMode(_ sender: UIButton)
    {
        Optimove.shared.stopTestMode()
    }
}
